napari/layers/shapes/_shapes_utils.py

```diff
@@ -732,6 +971,16 @@ def generate_2D_edge_meshes(path, closed=False, limit=3, bevel=False):
         Px3 array of the indices of the vertices that will form the
         triangles of the triangulation
     """
+    if closed and not contiguous and len(path) > 2:
+        # polygon — might include holes, we remove internal edges
+        return _combine_meshes(
+            [
+                generate_2D_edge_meshes(
+                    p, closed=closed, limit=limit, bevel=bevel, contiguous=True
+                )
+                for p in _remove_internal_edges(path)
+            ]
+        )
```

Comment:
If there is no hole, then calling `_remove_internal_edges` is expensive. However, we should know if it is required after `triangulate_face` call